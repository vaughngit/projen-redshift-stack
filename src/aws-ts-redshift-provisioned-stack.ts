//import {App, Stack, Tags, StackProps, CfnOutput,  Duration, RemovalPolicy, Fn, custom_resources as cr, CustomResource} from 'aws-cdk-lib';
import { Cluster, ClusterType } from '@aws-cdk/aws-redshift-alpha';
import { Stack, Tags, StackProps, CfnOutput, Duration, RemovalPolicy } from 'aws-cdk-lib';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
//import {aws_ssm as ssm } from 'aws-cdk-lib'
import { Construct } from 'constructs';
//import * as ec2 from 'aws-cdk-lib/aws-ec2';
//import * as fs from 'fs';
//import { aws_lambda as lambda } from 'aws-cdk-lib';
//import * as path from 'path';

export interface IStackProps extends StackProps {
  environment: string;
  costcenter: string;
  solution: string;
  appName: string;
  redShiftMasterUsername: string;
  // codecommit: codecommit.Repository;
}


export class AwsTsRedshiftStack extends Stack {
  constructor(scope: Construct, id: string, props: IStackProps) {
    super(scope, id, props);
    /*
    const vpc = new Vpc(this, `${props.appName}-vpc`, {
      vpcName: `${props.appName}-vpc`,
      maxAzs: 1
    });
     */

    //  const vpcId = ssm.StringParameter.valueFromLookup(this, '/ergon/dev/vpcId');
    // const vpc = Vpc.fromLookup(this, `${props.appName}-vpc`, {vpcId: vpcId.toString()})

    const vpc = Vpc.fromLookup(this, `${props.appName}-vpc`, { isDefault: true });

    /*
    const vpc = Vpc.fromVpcAttributes(this, `${props.appName}-vpc`, {
      vpcId: Fn.importValue('ergon:dev:VPCID:us-east-2').toString(),
      availabilityZones: [
        Fn.importValue(`ergon:PrivateSubnet1`),
        Fn.importValue(`ergon:PrivateSubnet2`)
      ]
    }) */


    //Redshift needs a role so that it can consume that role to download data from S3 bucket
    const redshiftDataBucket = new s3.Bucket(this, 'RedshiftDataBucket', {
      bucketName: `${props.appName}-${this.account}-${this.region}-redshift-data`,
      removalPolicy: RemovalPolicy.DESTROY,
      eventBridgeEnabled: true,
      autoDeleteObjects: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      //notificationsHandlerRole: RoleForS3ToInvokeLambda,
      //encryption: s3.BucketEncryption.KMS,
      //encryptionKey: encryptionKey,
      lifecycleRules: [
        {
          // ðŸ‘‡ optionally apply object name filtering
          // prefix: 'data/',
          abortIncompleteMultipartUploadAfter: Duration.days(30),
          expiration: Duration.days(365),
          transitions: [
            {
              storageClass: s3.StorageClass.INFREQUENT_ACCESS,
              transitionAfter: Duration.days(30),
            },
            {
              storageClass: s3.StorageClass.INTELLIGENT_TIERING,
              transitionAfter: Duration.days(60),
            },
            {
              storageClass: s3.StorageClass.GLACIER,
              transitionAfter: Duration.days(90),
            },
            {
              storageClass: s3.StorageClass.DEEP_ARCHIVE,
              transitionAfter: Duration.days(180),
            },
          ],
        },
      ],
    });


    const redShiftRole = new iam.Role(this, 'redShiftRole', {
      roleName: `${props.solution}-redShiftRole-${props.environment}`,
      description: 'Grants permission to redshift cluster.',
      assumedBy: new iam.ServicePrincipal('redshift.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('AWSGlueConsoleFullAccess'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonRedshiftFullAccess'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSageMakerFullAccess'),
      ],
    });
    redshiftDataBucket.grantRead(redShiftRole);

    // ðŸ‘‡ Create Trust Entity Policies:
    redShiftRole.assumeRolePolicy?.addStatements(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal('sagemaker.amazonaws.com')],
        actions: ['sts:AssumeRole'],
      }),
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal('events.amazonaws.com')],
        actions: ['sts:AssumeRole'],
      }),
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal('scheduler.redshift.amazonaws.com')],
        actions: ['sts:AssumeRole'],
      }),
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        principals: [new iam.AccountRootPrincipal()],
        actions: ['sts:AssumeRole'],
      }),
    );


    //create multi-node cluster with 2 compute nodes type DC2_LARGE (as default)
    //masterUser is created with password stored in secret manager (as default option)

    new Cluster(this, `${props.appName}-cluster-demo`, {
      clusterName: `${props.appName}-demo`,
      vpc: vpc,
      masterUser: {
        masterUsername: props.redShiftMasterUsername,
      },
      numberOfNodes: 2,
      clusterType: ClusterType.MULTI_NODE,
      removalPolicy: RemovalPolicy.DESTROY,
      roles: [redShiftRole],
      loggingProperties: {
        loggingBucket: redshiftDataBucket,
        loggingKeyPrefix: 'redshiftLogs/',
      },

    });


    /*
    //custom resource function:
    const LambdaPopulateSSMParameter = new lambda.Function(this, 'LambdaPopulateSSMParameter', {
      functionName: `${this.stackName}-PopulateSSMParameter`,
     code: lambda.Code.fromAsset(path.join(__dirname, "../sourceCode/lambda/LambdaPopulateSSMParameter/")),
     // memorySize: 1024,
      handler: "index.lambda_handler",
      timeout: Duration.seconds(60),
     // layers: [awssdk, datefnstz, nanopackage, uuidpackage],
      role: LambdaTwitterPermission,
      runtime: lambda.Runtime.PYTHON_3_8,
      environment: {
        TWITTER_MENTION_USER: TwitterMentionUser
      },
    });


    const provider = new cr.Provider(this, "Provider", {
      onEventHandler: LambdaPopulateSSMParameter,
      //role: LambdaTwitterPermission
    });

    //Custom Resource 1:
    new CustomResource(this, "CustomResource", {
      serviceToken: provider.serviceToken,
       properties: {
        BearerToken: BearerToken.secretArn,
        TwitterUser: TwitterUser.parameterName,
        tags:[  {Key: "service", Value: props.serviceName},{Key: "environment", Value: props.environment}, {Key: "solution", Value: props.solution},{Key: "costcenter", Value: props.costcenter}    ]

      },
    });


 */
    /*
    const clusterSg = cluster.connections.securityGroups[0];
    clusterSg.addIngressRule(clusterSg,  ec2.Port.allTcp(), "Allow internal access Redshift");

    const rsclient = new ec2.Instance(this, `${props.appName}-psql`, {
      instanceName: `${props.appName}-psql`,
      vpc: vpc,
      securityGroup: clusterSg,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.SMALL),
      machineImage: new ec2.AmazonLinuxImage({generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX_2}),
      role: new iam.Role(this, `${props.appName}-ec2-ssm`, {
        roleName: `${props.appName}-ec2-ssm`,
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [{managedPolicyArn: 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore'}]
      })
    });

    const userData = fs.readFileSync(path.join(__dirname, './user-data.sh'), 'utf8');
    rsclient.addUserData(userData);

 */
    Tags.of(this).add('solution', props.solution);
    Tags.of(this).add('environment', props.environment);
    Tags.of(this).add('costcenter', props.costcenter);

    new CfnOutput(this, 'ReShiftSqlWorkbenchRrl', { value: `https://${this.region}.console.aws.amazon.com/sqlworkbench/home?region=${this.region}#/client`, description: 'The RedShift SQL Workbench Url' });

    //https://us-east-2.console.aws.amazon.com/sqlworkbench/home?region=us-east-2#/client

  }
}
