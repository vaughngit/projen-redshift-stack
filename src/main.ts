import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
//import { AwsTsRedshiftStack } from './aws-ts-redshift-stack';
import { RedshiftServerlessStack } from './aws-ts-redshift-serverless-stack';

const solution = 'RedShift';
const environment = 'dev';
const costcenter = '12_1_12_9_20_8';
const aws_region = 'us-east-2';
const appName = 'redshift';
const desc = 'Redshift Demo Stack';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // define resources here...
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: aws_region || process.env.CDK_DEFAULT_REGION, //process.env.CDK_DEFAULT_REGION,
};

const app = new App();

//new MyStack(app, 'projen-redshift-stack-dev', { env: devEnv });
// new MyStack(app, 'projen-redshift-stack-prod', { env: prodEnv });

new RedshiftServerlessStack(app, 'serverless', {
  env: devEnv,
  stackName: `${solution}-serverless-${appName}`,
  redShiftMasterUsername: 'awsuser',
  rssnamespace: 'demoworkspaces',
  rssworkgroup: 'rssworkshop',
  appName,
  description: desc,
  solution,
  environment,
  costcenter,
  tags: {
    solution,
    environment,
    costcenter,
  },
});
/*
new AwsTsRedshiftStack(app, 'AwsTsRedshiftStack', {
  env: {account, region},
  stackName: `${solution}-${appName}`,
  redShiftMasterUsername: "awsuser",
  appName,
  description: desc,
  solution,
  environment,
  costcenter,
  tags: {
    solution,
    environment,
    costcenter
  },
});
 */


app.synth();