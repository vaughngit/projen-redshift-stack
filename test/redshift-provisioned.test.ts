import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
//import { MyStack } from '../src/main';
import { AwsTsRedshiftStack } from '../src/aws-ts-redshift-provisioned-stack';

const solution = 'RedShift';
const environment = 'dev';
const costcenter = 'tnc';
const aws_region = 'us-east-2';
const appName = 'redshift';
const desc = 'Redshift Demo Stack';
// for development, use account/region from cdk cli
const devEnv = {
  account: '796072252262' || process.env.CDK_DEFAULT_ACCOUNT,
  region: aws_region || process.env.CDK_DEFAULT_REGION, //process.env.CDK_DEFAULT_REGION,
};


test('Snapshot', () => {
  const app = new App();
  //const stack = new MyStack(app, 'test');
  const stack = new AwsTsRedshiftStack(app, 'AwsTsRedshiftStack', {
    env: devEnv,
    stackName: `${solution}-${appName}`,
    redShiftMasterUsername: 'awsuser',
    appName,
    description: desc,
    solution,
    environment,
    costcenter,
    tags: {
      solution,
      environment,
      costcenter,
    },
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});