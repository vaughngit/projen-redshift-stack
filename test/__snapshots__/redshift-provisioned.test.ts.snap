// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Description": "Redshift Demo Stack",
  "Outputs": {
    "ReShiftSqlWorkbenchRrl": {
      "Description": "The RedShift SQL Workbench Url",
      "Value": "https://us-east-2.console.aws.amazon.com/sqlworkbench/home?region=us-east-2#/client",
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": {
      "DependsOn": [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "import boto3  # type: ignore
import json
import logging
import urllib.request

s3 = boto3.client("s3")

EVENTBRIDGE_CONFIGURATION = 'EventBridgeConfiguration'

CONFIGURATION_TYPES = ["TopicConfigurations", "QueueConfigurations", "LambdaFunctionConfigurations"]

def handler(event: dict, context):
  response_status = "SUCCESS"
  error_message = ""
  try:
    props = event["ResourceProperties"]
    bucket = props["BucketName"]
    notification_configuration = props["NotificationConfiguration"]
    request_type = event["RequestType"]
    managed = props.get('Managed', 'true').lower() == 'true'
    stack_id = event['StackId']

    if managed:
      config = handle_managed(request_type, notification_configuration)
    else:
      config = handle_unmanaged(bucket, stack_id, request_type, notification_configuration)

    put_bucket_notification_configuration(bucket, config)
  except Exception as e:
    logging.exception("Failed to put bucket notification configuration")
    response_status = "FAILED"
    error_message = f"Error: {str(e)}. "
  finally:
    submit_response(event, context, response_status, error_message)

def handle_managed(request_type, notification_configuration):
  if request_type == 'Delete':
    return {}
  return notification_configuration

def handle_unmanaged(bucket, stack_id, request_type, notification_configuration):
  external_notifications = find_external_notifications(bucket, stack_id)

  if request_type == 'Delete':
    return external_notifications

  def with_id(notification):
    notification['Id'] = f"{stack_id}-{hash(json.dumps(notification, sort_keys=True))}"
    return notification

  notifications = {}
  for t in CONFIGURATION_TYPES:
    external = external_notifications.get(t, [])
    incoming = [with_id(n) for n in notification_configuration.get(t, [])]
    notifications[t] = external + incoming

  if EVENTBRIDGE_CONFIGURATION in notification_configuration:
    notifications[EVENTBRIDGE_CONFIGURATION] = notification_configuration[EVENTBRIDGE_CONFIGURATION]
  elif EVENTBRIDGE_CONFIGURATION in external_notifications:
    notifications[EVENTBRIDGE_CONFIGURATION] = external_notifications[EVENTBRIDGE_CONFIGURATION]

  return notifications

def find_external_notifications(bucket, stack_id):
  existing_notifications = get_bucket_notification_configuration(bucket)
  external_notifications = {}
  for t in CONFIGURATION_TYPES:
    external_notifications[t] = [n for n in existing_notifications.get(t, []) if not n['Id'].startswith(f"{stack_id}-")]

  if EVENTBRIDGE_CONFIGURATION in existing_notifications:
    external_notifications[EVENTBRIDGE_CONFIGURATION] = existing_notifications[EVENTBRIDGE_CONFIGURATION]

  return external_notifications

def get_bucket_notification_configuration(bucket):
  return s3.get_bucket_notification_configuration(Bucket=bucket)

def put_bucket_notification_configuration(bucket, notification_configuration):
  s3.put_bucket_notification_configuration(Bucket=bucket, NotificationConfiguration=notification_configuration)

def submit_response(event: dict, context, response_status: str, error_message: str):
  response_body = json.dumps(
    {
      "Status": response_status,
      "Reason": f"{error_message}See the details in CloudWatch Log Stream: {context.log_stream_name}",
      "PhysicalResourceId": event.get("PhysicalResourceId") or event["LogicalResourceId"],
      "StackId": event["StackId"],
      "RequestId": event["RequestId"],
      "LogicalResourceId": event["LogicalResourceId"],
      "NoEcho": False,
    }
  ).encode("utf-8")
  headers = {"content-type": "", "content-length": str(len(response_body))}
  try:
    req = urllib.request.Request(url=event["ResponseURL"], headers=headers, data=response_body, method="PUT")
    with urllib.request.urlopen(req) as response:
      print(response.read().decode("utf-8"))
    print("Status code: " + response.reason)
  except Exception as e:
      print("send(..) failed executing request.urlopen(..): " + str(e))
",
        },
        "Description": "AWS CloudFormation handler for "Custom::S3BucketNotifications" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": [
          {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789876-us-east-2",
          "S3Key": "3f51abb709b8e65167a45aeed02bab11540603d909005d7148230ba5ce6c74d7.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "RedshiftDataBucket8A6BF725",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RedshiftDataBucket8A6BF725": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "redshift-123456789876-us-east-2-redshift-data",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "AbortIncompleteMultipartUpload": {
                "DaysAfterInitiation": 30,
              },
              "ExpirationInDays": 365,
              "Status": "Enabled",
              "Transitions": [
                {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 30,
                },
                {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 60,
                },
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
                {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 180,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "RedshiftDataBucketAutoDeleteObjectsCustomResource3D64C675": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "RedshiftDataBucketPolicyC0B65F05",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "RedshiftDataBucket8A6BF725",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "RedshiftDataBucketNotificationsB125DF44": {
      "Properties": {
        "BucketName": {
          "Ref": "RedshiftDataBucket8A6BF725",
        },
        "Managed": true,
        "NotificationConfiguration": {
          "EventBridgeConfiguration": {},
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "RedshiftDataBucketPolicyC0B65F05": {
      "Properties": {
        "Bucket": {
          "Ref": "RedshiftDataBucket8A6BF725",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "RedshiftDataBucket8A6BF725",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RedshiftDataBucket8A6BF725",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "redshift.amazonaws.com",
              },
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RedshiftDataBucket8A6BF725",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
                {
                  "Fn::GetAtt": [
                    "RedshiftDataBucket8A6BF725",
                    "Arn",
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "redShiftRoleDefaultPolicy2ED2B019": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "RedshiftDataBucket8A6BF725",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "RedshiftDataBucket8A6BF725",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "redShiftRoleDefaultPolicy2ED2B019",
        "Roles": [
          {
            "Ref": "redShiftRoleEE0159B1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "redShiftRoleEE0159B1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "redshift.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "sagemaker.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.redshift.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::123456789876:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Grants permission to redshift cluster.",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonS3FullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSGlueConsoleFullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonRedshiftFullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSageMakerFullAccess",
              ],
            ],
          },
        ],
        "RoleName": "RedShift-redShiftRole-dev",
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "redshiftclusterdemo727C5D42": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AllowVersionUpgrade": true,
        "AutomatedSnapshotRetentionPeriod": 1,
        "ClusterIdentifier": "redshift-demo",
        "ClusterSubnetGroupName": {
          "Ref": "redshiftclusterdemoSubnets95CFDD46",
        },
        "ClusterType": "multi-node",
        "DBName": "default_db",
        "Encrypted": true,
        "IamRoles": [
          {
            "Fn::GetAtt": [
              "redShiftRoleEE0159B1",
              "Arn",
            ],
          },
        ],
        "LoggingProperties": {
          "BucketName": {
            "Ref": "RedshiftDataBucket8A6BF725",
          },
          "S3KeyPrefix": "redshiftLogs/",
        },
        "MasterUserPassword": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:",
              {
                "Ref": "redshiftclusterdemoSecretD124EDA1",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:",
              {
                "Ref": "redshiftclusterdemoSecretD124EDA1",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "NodeType": "dc2.large",
        "NumberOfNodes": 2,
        "PubliclyAccessible": false,
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
        "VpcSecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "redshiftclusterdemoSecurityGroup2040A8C0",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::Redshift::Cluster",
      "UpdateReplacePolicy": "Delete",
    },
    "redshiftclusterdemoSecretAttachmentA4AFB5DF": {
      "Properties": {
        "SecretId": {
          "Ref": "redshiftclusterdemoSecretD124EDA1",
        },
        "TargetId": {
          "Ref": "redshiftclusterdemo727C5D42",
        },
        "TargetType": "AWS::Redshift::Cluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "redshiftclusterdemoSecretD124EDA1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "GenerateSecretString": {
          "ExcludeCharacters": ""@/\\ '",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{"username":"awsuser"}",
        },
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "redshiftclusterdemoSecurityGroup2040A8C0": {
      "Properties": {
        "GroupDescription": "Redshift security group",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "redshiftclusterdemoSubnets95CFDD46": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "Subnets for redshift-cluster-demo Redshift cluster",
        "SubnetIds": [
          "p-12345",
          "p-67890",
        ],
        "Tags": [
          {
            "Key": "costcenter",
            "Value": "tnc",
          },
          {
            "Key": "environment",
            "Value": "dev",
          },
          {
            "Key": "solution",
            "Value": "RedShift",
          },
        ],
      },
      "Type": "AWS::Redshift::ClusterSubnetGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
